libs/src/services/discoveryProvider/DiscoveryProviderSelection.js

```diff
@@ -0,0 +1,187 @@
+  REGRESSED_MODE_TIMEOUT
+} = require('./constants')
+const semver = require('semver')
+
+let localStorage
+if (typeof window === 'undefined' || window === null) {
+  const LocalStorage = require('node-localstorage').LocalStorage
+  localStorage = new LocalStorage('./local-storage')
+} else {
+  localStorage = window.localStorage
+}
+
+class DiscoveryProviderSelection extends ServiceSelection {
+  constructor (config, ethContracts) {
+    super({
+      /**
+       * Gets the "current" expected service version as well as
+       * the list of registered providers from chain
+       */
+      getServices: async () => {
+        this.currentVersion = await ethContracts.getCurrentVersion(DISCOVERY_SERVICE_NAME)
+        return this.ethContracts.getServiceProviderList(DISCOVERY_SERVICE_NAME)
+      },
+      ...config
+    })
+    this.ethContracts = ethContracts
+    this.currentVersion = null
+
+    // Whether or not we are running in `regressed` mode, meaning we were
+    // unable to select a discovery provider that was up-to-date. Clients may
+    // want to consider blocking writes.
+    this._regressedMode = false
+  }
+
+  /** Retrieves a cached discovery provider from localstorage */
+  getCached () {
+    if (localStorage) {
+      const discProvTimestamp = localStorage.getItem(DISCOVERY_PROVIDER_TIMESTAMP)
+      if (discProvTimestamp) {
+        const { endpoint: latestEndpoint, timestamp } = JSON.parse(discProvTimestamp)
+        const inWhitelist = !this.whitelist || this.whitelist.has(latestEndpoint)
+        const isExpired = (Date.now() - timestamp) > DISCOVERY_PROVIDER_RESELECT_TIMEOUT
+        if (!inWhitelist || isExpired) {
+          this.clearCached()
+        } else {
+          return latestEndpoint
+        }
+      }
+    }
+    return null
+  }
+
+  /** Clears any cached discovery provider from localstorage */
+  clearCached () {
+    if (localStorage) {
+      localStorage.removeItem(DISCOVERY_PROVIDER_TIMESTAMP)
+    }
+  }
+
+  /** Sets a cached discovery provider in localstorage */
+  setCached (endpoint) {
+    localStorage.setItem(DISCOVERY_PROVIDER_TIMESTAMP, JSON.stringify({ endpoint, timestamp: Date.now() }))
+  }
+
+  /** Allows the selection take a shortcut if there's a cached provider */
+  shortcircuit () {
+    return this.getCached()
+  }
+
+  async select () {
+    const endpoint = await super.select()
+    this.setCached(endpoint)
+    return endpoint
+  }
+
+  /**
+   * Checks whether a given response is healthy:
+   * - Not behind in blocks
+   * - 200 response
+   * - Current version
+   *
+   * Other responses are collected in `this.backups` if
+   * - Behind by only a patch version
+   *
+   * @param {Response} response axios response
+   * @param {{ [key: string]: string}} urlMap health check urls mapped to their cannonical url
+   * e.g. https://discoveryprovider.audius.co/health_check => https://discoveryprovider.audius.co
+   */
+  isHealthy (response, urlMap) {
```

Comment:
overriding the parent class 'isHealthy', so we can make different criteria for cnode vs disc prov - very cool